src/winhello/WindowsHello.cpp

```diff
@@ -0,0 +1,233 @@
+/*
+ *  Copyright (C) 2022 KeePassXC Team <team@keepassxc.org>
+ *  Copyright (C) 2022 Thomas Hobson (HexF) <thomas@hexf.me>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "WindowsHello.h"
+
+#include <QApplication>
+#include <QCoreApplication>
+#include <QHash>
+#include <QWidget>
+
+#include <windows.h>
+
+#ifndef MS_NGC_KEY_STORAGE_PROVIDER
+#define MS_NGC_KEY_STORAGE_PROVIDER L"Microsoft Passport Key Storage Provider" // MinGW doesn't provide this
+#endif
+#ifndef NCRYPT_PIN_CACHE_IS_GESTURE_REQUIRED_PROPERTY
+#define NCRYPT_PIN_CACHE_IS_GESTURE_REQUIRED_PROPERTY L"PinCacheIsGestureRequired"
+#endif
+
+using namespace WindowsHello;
+
+namespace
+{
+    QHash<QString, QByteArray> g_encryptedMasterKeys;
+
+    int g_winHelloAvailable = WINDOWSHELLO_UNDEFINED;
+
+    bool getNCryptHandles(NCRYPT_PROV_HANDLE* hProv, NCRYPT_KEY_HANDLE* hKey)
+    {
+        if (NCryptOpenStorageProvider(hProv, MS_NGC_KEY_STORAGE_PROVIDER, 0) != ERROR_SUCCESS) {
+            g_winHelloAvailable = WINDOWSHELLO_NOT_AVAILABLE;
+            return false;
+        }
+
+        NCryptKeyName* keyname = nullptr;
+        PVOID pos = nullptr;
+
+        // There is either 0 or 1 keys in here - we need to get the name so we just enumerate the store.
+        // If we get an error, its empty and can exit
+        if (NCryptEnumKeys(*hProv, NULL, &keyname, &pos, 0) != ERROR_SUCCESS) {
+            g_winHelloAvailable = WINDOWSHELLO_NOT_AVAILABLE;
+            return false;
+        }
+
+        if (NCryptOpenKey(*hProv, hKey, keyname->pszName, keyname->dwLegacyKeySpec, 0) != ERROR_SUCCESS) {
+            NCryptFreeBuffer(keyname);
+
+            g_winHelloAvailable = WINDOWSHELLO_NOT_AVAILABLE;
+            return false;
+        }
+
+        NCryptFreeBuffer(keyname);
+
+        QString promptMessage =
+            QCoreApplication::translate("DatabaseOpenDialog", "authenticate to access the database");
+        wchar_t* prompt = new wchar_t[promptMessage.length() * 2];
+        promptMessage.toWCharArray(prompt);
+
+        if (NCryptSetProperty(*hKey,
+                              NCRYPT_USE_CONTEXT_PROPERTY,
+                              reinterpret_cast<PBYTE>(prompt),
+                              (promptMessage.length() + 1) * 2,
```

Comment:
Here you should use the return value of `toWCharArray() + sizeof(wchar_t)`  instead.

Reply:
ToWCharArray does not append a null terminator https://doc.qt.io/qt-5/qstring.html#toWCharArray

Reply:
It does not, but `NCryptSetProperty()` expects a null-terminated string (+ the size) for some reason.